bootstrap/secret/secure.go

```diff
@@ -0,0 +1,180 @@
+/*******************************************************************************
+ * Copyright 2018 Dell Inc.
+ * Copyright 2020 Intel Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package secret
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/interfaces"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	"github.com/edgexfoundry/go-mod-secrets/pkg/token/authtokenloader"
+	"github.com/edgexfoundry/go-mod-secrets/secrets"
+
+	"sync"
+	"time"
+)
+
+// Provider implements the SecretProvider interface
+type SecureProvider struct {
+	secretClient  secrets.SecretClient
+	lc            logger.LoggingClient
+	loader        authtokenloader.AuthTokenLoader
+	configuration interfaces.Configuration
+	secretsCache  map[string]map[string]string // secret's path, key, value
+	cacheMutex    *sync.Mutex
+	lastUpdated   time.Time
+}
+
+// NewSecureProvider creates & initializes Provider instance for secure secrets.
+func NewSecureProvider(config interfaces.Configuration, lc logger.LoggingClient, loader authtokenloader.AuthTokenLoader) *SecureProvider {
+	provider := &SecureProvider{
+		configuration: config,
+		lc:            lc,
+		loader:        loader,
+		secretsCache:  make(map[string]map[string]string),
+		cacheMutex:    &sync.Mutex{},
+		lastUpdated:   time.Now(),
+	}
+	return provider
+}
+
+// SetClient sets the secret client that is used to access the secure secrets
+func (p *SecureProvider) SetClient(client secrets.SecretClient) {
+	p.secretClient = client
+}
+
+// GetSecrets retrieves secrets from a secret store.
+// path specifies the type or location of the secrets to retrieve.
+// keys specifies the secrets which to retrieve. If no keys are provided then all the keys associated with the
+// specified path will be returned.
+func (p *SecureProvider) GetSecrets(path string, keys ...string) (map[string]string, error) {
+	if cachedSecrets := p.getSecretsCache(path, keys...); cachedSecrets != nil {
+		return cachedSecrets, nil
+	}
+
+	if p.secretClient == nil {
+		return nil, errors.New("can't get secret(p), secret client is not properly initialized")
```

Comment:
what is `secret(p)` ?  you meant cannot get secrets from `path`?

Reply:
Not sure. Think that was existing code. Will fix up the message :-)